adoc/chapters/programming_interface.adoc

```diff
@@ -19401,481 +19401,697 @@ that all of the built-in functions fulfill the same requirements for both
+----
+genfloat fmax(genfloat x, genfloat y)
+----
+
+The generic type [code]#genfloat# is a stand-in for a set of a actual type
+names: [code]#float#, [code]#double#, etc.  To construct the actual overloads
+that are available, replace the generic type name with the name of each of
+these actual types:
+
+[source]
+----
+float fmax(float x, float y)
+double fmax(double x, double y)
+template<int N> vec<float,N> fmax(vec<float,N> x, vec<float,N> y)
+template<size_t M> marray<float,M> fmax(marray<float,M> x, marray<float,M> y)
+// etc.
+----
+
+When determining the set of overloads that are available, use the following
+rules:
+
+* If the specification of a function has several occurrences of the same
+  generic type name, replace each occurrence with the same actual type.  For
+  example, the [code]#fmax# example above has three occurrences of
+  [code]#genfloat#: one for the return type and one for each parameter type.
+  The available overloads include only the cases where both parameters have
+  the same type, and the return type always matches the type of the parameters.
+
+* Some of the generic type names can be replaced with a templated type, where
+  the identifiers [code]#N#, [code]#M#, [code]#Space#, and [code]#IsDecorated#
+  represent template parameters.  When this happens, the overload is a
+  templated function, as shown above with the [code]#fmax# example.  Only a few
+  combinations of template parameters are possible, and the following list
+  tells the order and type of the template parameters for each possible
+  combination:
+
+** If there is a single template parameter [code]#N# (number of dimensions of
+   the [code]#vec# type):
++
+[source]
+----
+template<int N>
+----
+
+** If there is a single template parameter [code]#M# (number of dimensions of
+   the [code]#marray# type):
++
+[source]
+----
+template<size_t M>
+----
+
+** If there are two template parameters [code]#Space# and [code]#IsDecorated#
+   (address space and decoration for the [code]#multi_ptr# type):
++
+[source]
+----
+template<access::address_space Space, access::decorated IsDecorated>
+----
 
+** If there are three template parameters [code]#N#, [code]#Space# and
+   [code]#IsDecorated#:
++
+[source]
+----
+template<int N, access::address_space Space, access::decorated IsDecorated>
+----
+
+** If there are three template parameters [code]#M#, [code]#Space# and
+   [code]#IsDecorated#:
++
+[source]
+----
+template<size_t M, access::address_space Space, access::decorated IsDecorated>
+----
+
+* Unless specified otherwise, the template parameter [code]#N# is constrained
+  to the values 2, 3, 4, 8, or 16.
+
+* Unless specified otherwise, the template parameter [code]#M# is constrained
+  to a positive value.
+
+* Unless specified otherwise, the template parameter [code]#Space# is
+  constrained to the values [code]#access::address_space::global_space#,
+  [code]#access::address_space::local_space#, or
+  [code]#access::address_space::private_space#.
```

Comment:
Reading this again this morning... Why aren't these functions available for generic pointers?  Pointers are only ever used for return types, so it seems odd to limit the available address spaces here.

It would make sense to keep a note that `Space` cannot be `constant_space`, but we could remove that in SYCL-Next (since `constant_space` is deprecated).

Reply:
I noticed this too, and I was thinking of creating an issue for this.  The spec had the same limitation before my PR, and I didn't want to expand the scope of this PR to include new additions.  In addition to adding `generic_space`, I think we should consider adding new overloads that take raw pointers.

Reply:
I have addressed this in https://github.com/KhronosGroup/SYCL-Docs/pull/432 which is rebased on this PR.